Issue #13409 ensure classpath Resource conversion uses full path

janbartel opened the pull request:
closes #13409

The problem described in #13409 has 2 parts to it:

1.  #11092 converted all classpath jars into `MountedPathResource` instead of `PathResource`
2.  after the change in #10889 to use non-ephemeral `Resources` to refer to the results of scanning for `web-fragment.xml`, `*.tld` and `META-INF/resources` (so that we avoid keeping mount points around for jars that don't actually contain any of these things), whenever we compare `Resources` we must ensure that their `path`s are the same.

For example, when running with a command line like:

> java -cp target/dependency/*:target/application-1.0.jar  some.Main

and `target/dependency/library-1.0.jar` contains a `web-fragment.xml` we see the following:

The container `Resource` is a `MountedPathResource` and has:

```
path = /
uri = jar:file:///home/janb/src/tmp/application/target/dependency/library-1.0.jar!/
```
The `Resource` of the jar known to contain the `web-fragment.xml` is a `PathResource` and has:

```
path=/home/janb/src/tmp/application/target/dependency/library-1.0.jar
uri=file:///home/janb/src/tmp/application/target/dependency/library-1.0.jar
```
As can be seen, there are 2 problems:

1 A comparison of these 2 resources fails because `MountedPathResource` is not a `PathResource` 
2 Even if they were both `PathResources`, the paths are different 

Reverting the change to `MetaInfConfiguration` from #11092, we now have both `Resources` as `PathResource`, but one resource has a relative path whilst the other is fully qualified:

The container `Resource`  has:

```
path=target/dependency/library-1.0.jar
uri=file:///home/janb/src/tmp/application/target/dependency/library-1.0.jar
```

The `Resource` that represents the jar with fragment has:

```
path=/home/janb/src/tmp/application/target/dependency/library-1.0.jar
uri=file:///home/janb/src/tmp/application/target/dependency/library-1.0.jar
```

This PR reverts the change from #11092, but also ensures that _both_ the container `Resource` and jar containing the fragment `Resource` have fully qualified paths.

@joakime is there a way to modify your change from #11092 to keep some of the goodness, but ensure that it doesn't always create `MountedPathResource`? We definitely don't want to always do the mount, because we only want to mount if we are going to be keeping the resource around because it has something useful in it (ie see the changes we made in #10889). We do need to retain the fix in this PR to ensure that the paths are fully qualified.

janbartel committed:
Issue #13409 ensure classpath Resource conversion uses full path

joakime commented on jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/MetaInfConfiguration.java:
```diff
@@ -168,10 +169,17 @@ public void findAndFilterContainerPaths(final WebAppContext context) throws Exce
+            //fully qualified path. In MetaData.resolve() we compare this
+            //fully qualified path with the container Resources, and thus both
+            //must be fully qualified for a match (as per PathResource.equals() method).
+            Stream.of(classPath.split(File.pathSeparator))
```
This should be fixed in resourceFactory.split() instead, as that's used from multiple place that need this same logic.  (eg: WebAppContext.setExtraClassPath())

joakime reviewed (changes requested):
The fix should be in ResourceFactory.split()

janbartel committed:
Move classpath split to ResourceFactory.

Ensure that jar:file urls are (optionally) unwrapped.
Ensure that relative classpaths are absolute.
Ensure that ResourceFactory.split does not result in MountedPathResources.

janbartel committed:
Fix tests

janbartel committed:
Fix test

janbartel committed:
Fix test

joakime commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/CombinedResourceTest.java:
```diff
@@ -455,6 +457,15 @@ public void testUserSpaceConfigurationWithGlob() throws Exception
+        {
+            Resource resource = i.next();
+            if (resource.exists() && !resource.isDirectory() && FileID.isLibArchive(resource.getName()))
+                i.set(resourceFactory.newResource(URIUtil.toJarFileUri(resource.getURI())));
```
This would eliminate the previous Resource that might need to be cleaned up by a future the ResourceFactory close (this looks like a Resource leak).

janbartel replied:
You always have to be aware that Resources you obtain from a ResourceFactory have to be cleaned up.

joakime commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceFactoryTest.java:
```diff
@@ -311,33 +314,59 @@ public void testNewResource(String inputRaw, String expectedUri)
             String input = testJar.toUri().toASCIIString();
             List<Resource> resources = resourceFactory.split(input);
-            String expected = URIUtil.toJarFileUri(testJar.toUri()).toASCIIString();
+            String expected = testJar.toUri().toASCIIString();
```
This is the same value as `input` now.
Seems wrong, as the result of `ResourceFactory.split()` should be the mounted JAR.

janbartel replied:
We cannot let `ResourceFactory.split()` do any jar mounting, otherwise we wind up with system resources being consumed for jars that ultimately don't contain anything that we are interested in.  That was the problem that we fixed in #10889. It can only be the caller of `split()` that can decide if it needs to mount in order to use the individual resources.

joakime commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceFactoryTest.java:
```diff
@@ -311,33 +314,59 @@ public void testNewResource(String inputRaw, String expectedUri)
             String input = testJar.toString();
             List<Resource> resources = resourceFactory.split(input);
-            String expected = URIUtil.toJarFileUri(testJar.toUri()).toASCIIString();
+            String expected = testJar.toUri().toASCIIString();
```
Ditto (from above)

joakime commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceFactoryTest.java:
```diff
@@ -311,33 +314,59 @@ public void testNewResource(String inputRaw, String expectedUri)
+    {
+        try (ResourceFactory.Closeable resourceFactory = ResourceFactory.closeable())
+        {
+            String config = String.format("%s,%s,%s,%s,%s,%s", "dir/a", "foo/b", "bar/c", "jar:file:///foo/bar.jar!/", "file:///foo/bar.jar", "/foo/bar.jar");
```
All of these would need to exist for this test to be valid.

janbartel replied:
Not really, if you read further in the test, we're just testing certain things about how it is converted, not that they exist. The comments explain it.

janbartel committed:
Fix ee8 test

joakime reviewed (changes requested):
Some thoughts

janbartel committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-13409-fully-qualified-Resource

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java:
```diff
@@ -499,25 +505,40 @@ default List<Resource> split(String str, String delims)
+                        // that might be unnecessary - the caller should always decide whether to mount
+                        URI uri = new URI(reference);
+                        if (unwrap)
+                            list.add(newResource(URIUtil.unwrapContainer(uri)));
```
Should we throw here if the path inside the jar is not / ?

joakime replied:
Nah, that's a valid use case IMO

janbartel replied:
Maybe, but not necessary???

joakime replied:
This isn't a restriction we need (right now)

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java:
```diff
@@ -499,25 +505,40 @@ default List<Resource> split(String str, String delims)
+                    {
+                        Path p = Paths.get(reference);
+                        if (!p.isAbsolute() && LOG.isDebugEnabled())
+                            LOG.warn("Non-absolute path: {}", reference);
```
throw rather than warn?

janbartel replied:
I think it's caller dependent - `MetaInfConfiguration` will have a problem if it's not an absolute path, but other callers may be ok with that.

joakime replied:
It should always return absolute Resource paths IMO.
Since this method returns `Resource`, there's no real useful way to work with relative paths in any way other than to convert it to absolute first.

How about ...

``` java
   Path p = Path.of(reference).toAbsolutePath(); // enforce absolute path.
```

The `Path.of(String)` will happily accept absolute, and relative strings.
We don't have to worry about URI strings, as that's a different code path protected by the `if (URIUtil.hasScheme(reference))` earlier.

The `Path.toAbsolutePath()` will just return the path if it's already absolute, but fix it if isn't absolute.

```
jshell> var p1 = Path.of("12.0.x/jetty-core/pom.xml")
p1 ==> 12.0.x/jetty-core/pom.xml

jshell> p1.toAbsolutePath()
$2 ==> /home/joakim/code/jetty/12.0.x/jetty-core/pom.xml

jshell> var p2 = Path.of("/home/joakim/code/jetty/9.4.x/jetty-util/src/main/config/modules/console-capture.mod")
p2 ==> /home/joakim/code/jetty/9.4.x/jetty-util/src/main ... odules/console-capture.mod

jshell> p2.toAbsolutePath()
$4 ==> /home/joakim/code/jetty/9.4.x/jetty-util/src/main/config/modules/console-capture.mod
```

janbartel replied:
If you look at the line below the `throw` you will see that we always convert to absolute path, so I think your concern is already addressed.

gregw commented on jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/MetaInfConfiguration.java:
```diff
@@ -168,10 +168,10 @@ public void findAndFilterContainerPaths(final WebAppContext context) throws Exce
         if (classPath != null)
         {
-            resourceFactory.split(classPath, File.pathSeparator)
+            resourceFactory.split(classPath, File.pathSeparator, true)
```
maybe a comment here on why you are passing true

gregw commented on jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java:
```diff
@@ -1259,7 +1260,7 @@ public List<Resource> getExtraClasspath()
     public void setExtraClasspath(String extraClasspath)
     {
-        setExtraClasspath(getResourceFactory().split(extraClasspath));
+        setExtraClasspath(getResourceFactory().split(extraClasspath, File.pathSeparator, true));
```
ditto

gregw commented on jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/MetaInfConfiguration.java:
```diff
@@ -170,10 +170,10 @@ public void findAndFilterContainerPaths(final WebAppContext context) throws Exce
         if (classPath != null)
         {
-            resourceFactory.split(classPath, File.pathSeparator)
+            resourceFactory.split(classPath, File.pathSeparator, true)
```
ditto

gregw commented on jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppContext.java:
```diff
@@ -1267,7 +1267,7 @@ public List<Resource> getExtraClasspath()
     public void setExtraClasspath(String extraClasspath) throws IOException
     {
-        setExtraClasspath(getResourceFactory().split(extraClasspath));
+        setExtraClasspath(getResourceFactory().split(extraClasspath, File.pathSeparator, true));
```
ditto

gregw reviewed (dismissed):
A couple of niggles but otherwise OK

janbartel committed:
Changes after review

janbartel committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-13409-fully-qualified-Resource

timja commented on jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java:
```diff
@@ -1259,7 +1260,10 @@ public List<Resource> getExtraClasspath()
+        //convert classpath into Resources. Any jar:file: references will be
+        //unwrapped, because we need the location of the file, not the contents
+        //of the file
+        setExtraClasspath(getResourceFactory().split(extraClasspath, File.pathSeparator, true));
```
Caused https://github.com/jetty/jetty.project/issues/13698
